0.data_analysis/utils/CorrelateData.py

```diff
@@ -0,0 +1,174 @@
+from collections import defaultdict
+from itertools import combinations, product
+
+import pandas as pd
+
+
+class CorrelateData:
+    """
+    Compute inter or intra group correlations between objects.
+    """
+
+    def __init__(self):
+        pass
+
+    def save_params(self, _groups, _cols, _drop_cols, _corrs):
+        """
+        Parameters
+        ----------
+        _groups: List(Tuple) or Tuple(List) or List or Tuple
+            The values for each group (groups 0 and 1) for each column to save.
+
+        _cols: List
+            Columns to save for each of the two groups.
+
+        _drop_cols: List
+            Columns to not save.
+
+        _corrs: Defaultdict(List)
+            Corrlelation data storing correlations and column values for each data pair.
+
+        Returns
+        -------
+        _corrs: Defaultdict(List)
+            Corrlelation data storing correlations and column values for each data pair.
+        """
+
+        # Ensure the column value can be indexed if only one column is saved for the group for compatibility
+        if (not isinstance(_groups[0], list)) and (not isinstance(_groups[0], tuple)):
+            _groups = list(_groups)
+            _groups[0], _groups[1] = [_groups[0]], [_groups[1]]
+
+        # Iterate through each column and save the value for that group and column
+        for idx, col in enumerate(_cols):
+            if col in _drop_cols:
+                continue
+            _corrs[f"{col}__group0"].append(_groups[0][idx])
+            _corrs[f"{col}__group1"].append(_groups[0][idx])
+
+        return _corrs
+
+    def inter_correlations(self, _df, _posthoc_group_cols, _feat_cols, _antehoc_group_cols, _drop_cols = []):
+        """
+        Computes correlations between two objects using both post hoc and ante groups.
+        This is accomplished by computing correlations between objects in different ante groups (inter correlations) and between all possible post hoc correlation comparisons between those groups.
+
+        Parameters
+        ----------
+        _df: pd.DataFrame
+            Contains the features and group columns to use for correlating objects.
+
+        _posthoc_group_cols: List(Objects)
+            Columns to groupby after organizing by _antehoc_group_cols.
+
+        _feat_col: List
+            Feature columns.
+
+        _posthoc_group_cols: List(Objects)
```

Comment:
```suggestion
        _antehoc_group_cols: List(Objects)
```
Looks like this parameter was mis-named.

Reply:
Good catch, changed